src/genai/extensions/lm_eval/model.py

```diff
@@ -0,0 +1,305 @@
+import json
+from collections import defaultdict
+from typing import Any, Iterator, NamedTuple, Optional, Type, cast
+
+from genai import Client, Credentials
+from genai.schema import (
+    BaseTokens,
+    DecodingMethod,
+    TextGenerationParameters,
+    TextGenerationReturnOptions,
+    TextTokenizationParameters,
+    TextTokenizationReturnOptions,
+)
+from genai.text.generation import CreateExecutionOptions as TextGenerationExecutionOptions
+from genai.text.tokenization import CreateExecutionOptions as TokenizationExecutionOptions
+
+try:
+    import lm_eval.utils
+    from lm_eval.api.instance import Instance
+    from lm_eval.api.model import LM
+    from lm_eval.api.registry import register_model
+    from lm_eval.models.utils import Grouper
+except ImportError:
+    raise ImportError("Could not import lm_eval: Please install ibm-generative-ai[lm-eval] extension.")  # noqa: B904
+try:
+    from tqdm import tqdm
+except ImportError:
+    raise ImportError("Could not import tqdm: Please install ibm-generative-ai[lm-eval] extension.")  # noqa: B904
+
+
+class LogLikelihoodResult(NamedTuple):
+    log_likelihood: float
+    is_greedy: bool
+
+
+def initialize_model():
+    pass  # model is registered by importing this module
+
+
+@register_model("ibm_genai")
+class IBMGenAILMEval(LM):
+    """
+    Implementation of LM model interface for evaluating GenAI model with the lm_eval framework.
+
+    See https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/model_guide.md for reference.
+    """
+
+    DEFAULT_TOKENIZATION_EXECUTION_OPTIONS = TokenizationExecutionOptions(
+        batch_size=100,
+        concurrency_limit=5,
+    )
+    DEFAULT_GENERATION_EXECUTION_OPTIONS = TextGenerationExecutionOptions()
+    DEFAULT_NUM_RETRIES = 6  # Increased number of retries for robustness, because evaluation typically runs for hours
+
+    @classmethod
+    def create_from_arg_string(
+        cls: Type["IBMGenAILMEval"],
+        arg_string: str,
+        additional_config: Optional[dict] = None,
+    ) -> "IBMGenAILMEval":
+        """Allow the user to specify model parameters (TextGenerationParameters) in CLI arguments."""
+        additional_config = {} if additional_config is None else additional_config
+        args = lm_eval.utils.simple_parse_args_string(arg_string)
+        model_id = args.pop("model_id", None)
+        if model_id is None:
+            raise ValueError("'model_id' is required, please pass it in 'model_args'")
+        parameters = TextGenerationParameters.model_validate(args)
+        return cls(client=Client(credentials=Credentials.from_env()), model_id=model_id, parameters=parameters)
+
+    def __init__(
+        self,
+        client: Optional[Client] = None,
+        model_id: Optional[str] = None,
+        parameters: Optional[TextGenerationParameters] = None,
+        show_progressbar: Optional[bool] = True,
```

Comment:
Can you move this option (`show_progressbar`) to concrete methods?

Reply:
The methods are typically not called by user, it's expected that user will initialize the model and pass it to `evaluate` or `simple_evaluate` from `lm_eval`